src/v/cluster/members_backend.cc

```diff
@@ -43,6 +44,183 @@
+ *                                             ___
+ *                                             ╲
+ *                                  (T - R) +  ╱    R
+ *                                             ‾‾‾
+ *                                            n = 1
+ *                           e    = ─────────────────
+ *                            max         T ⋅ N
+ *
+ * normalized error is equal to:
+ *
+ *                                 e
+ *                           e = ────
+ *                               e
+ *                                max
+ *
+ *
+ * Since the improvement depends on the number of replicas moved in a single
+ * batch we need to calculate the improvement that would moving a single
+ * partition replica introduce. Since single move requested by a balancer moves
+ * partition from node A->B (where A has more than requested number of replicas
+ * per nd B should have less than requested number of replicas) we can calculate
+ * the error improvement introduced by single move as:
+ *
+ *                                     2
+ *                           e     = ─────
+ *                            step   N ⋅ T
+ *
+ * Single step error must be normalized as well. We stop if the improvement is
+ * less than the equivalent of the improvement made by 1/10th of the number of
+ * moves in a batch (or 1 move if that is bigger)
+ **/
+unevenness_error_info calculate_unevenness_error(
+  const partition_allocator& allocator, partition_allocation_domain domain) {
+    const auto node_cnt = allocator.state().available_nodes();
+
+    const auto node_replicas = calculate_replicas_per_node(allocator, domain);
+    const auto total_replicas = calculate_total_replicas(node_replicas);
+
+    if (total_replicas == 0) {
+        return {.e = 0.0, .e_step = 0.0};
+    }
+
+    const auto target_replicas_per_node = total_replicas
+                                          / allocator.state().available_nodes();
+    // max error is an error calculated when all replicas are allocated on
+    // the same node
+    double max_err = (total_replicas - target_replicas_per_node)
+                     + target_replicas_per_node * (node_cnt - 1);
+    // divide by total replicas and node count to make the error independent
+    // from number of nodes and number of replicas.
+    max_err /= static_cast<double>(total_replicas);
+    max_err /= static_cast<double>(node_cnt);
+
+    double err = 0;
+    for (auto& [id, allocation_info] : node_replicas) {
+        double diff = static_cast<double>(target_replicas_per_node)
+                      - static_cast<double>(allocation_info.allocated_replicas);
+
+        vlog(
+          clusterlog.trace,
+          "node {} has {} replicas allocated in domain {}, requested replicas "
+          "per node "
+          "{}, difference: {}",
+          id,
+          allocation_info.allocated_replicas,
+          domain,
+          target_replicas_per_node,
+          diff);
+        err += std::abs(diff);
```

Comment:
side note: squaring the diff here is probably better because if we have, say, node partition counts 0, 2 (with the target count of 3), this is worse than 1, 1 (and the `least_allocated` constraint will prefer the latter), even though they have the same error in the L1 metric.